google-cloud-storage/acceptance/storage/bucket_post_object_test.rb

```diff
@@ -0,0 +1,90 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require "storage_helper"
+require 'net/http'
+require 'uri'
+
+describe Google::Cloud::Storage::Bucket, :post_object, :storage do
+  let(:bucket_name) { $bucket_names.first }
+  let :bucket do
+    storage.bucket(bucket_name) ||
+    safe_gcs_execute { storage.create_bucket(bucket_name) }
+  end
+  let(:uri) { URI.parse Google::Cloud::Storage::GOOGLEAPIS_URL }
+  let(:data_file) { "logo.jpg" }
+  let(:data) { File.expand_path("../data/#{data_file}", __dir__) }
+  let :policy do
+    {
+      expiration: (Time.now + 600).iso8601,
+      conditions: [
+        ["starts-with", "$key", ""]
+      ]
+    }
+  end
+
+  it "generates a signed post object" do
+    file_name = "logo-#{SecureRandom.hex(4).downcase}.jpg"
+
+    bucket.file(file_name).must_be :nil?
+
+    signed_post = bucket.post_object file_name, policy: policy
+
+    http = Net::HTTP.new(uri.host, uri.port)
+    http.use_ssl = true
+    request = Net::HTTP::Post.new(uri.request_uri)
+
+    form_data = [
+      ['file', File.open(data)],
+      ['key', signed_post.fields[:key]],
+      ['GoogleAccessId', signed_post.fields[:GoogleAccessId]],
+      ['policy', signed_post.fields[:policy]],
+      ['signature', signed_post.fields[:signature]]
+    ]
+    request.set_form form_data, 'multipart/form-data'
+
+    response = http.request(request)
+
+    response.code.must_equal "204"
+    bucket.file(file_name).wont_be :nil?
+  end
+
+  it "generates a signed post object with special variable key ${filename}" do
+    # "You can also use the ${filename} variable if a user is providing a file name."
+    #  https://cloud.google.com/storage/docs/xml-api/post-object
+    special_key = "${filename}"
+
+    bucket.file(data_file).must_be :nil?
+
+    signed_post = bucket.post_object special_key, policy: policy
+
+    http = Net::HTTP.new(uri.host, uri.port)
+    http.use_ssl = true
+    request = Net::HTTP::Post.new(uri.request_uri)
+
+    form_data = [
+      ['file', File.open(data)],
+      ['key', signed_post.fields[:key]],
```

Comment:
I was suspecting this to be a new name that didn't match `${filename}`, so it would imply the name can be added in the subsequent request.

Reply:
As I discovered in my manual testing yesterday (form below), if the `key` field is actually set to `[BUCKET]/${filename}` (literally `${filename}`) and is `hidden`, the user of the form only has to choose a file, and the object created in GCS will have the same name as the uploaded file, without requiring the `key` field also to be set by the user.

GCS documentation for this feature is minimal, but S3 [states](https://docs.aws.amazon.com/sdkforruby/api/Aws/S3/PresignedPost.html):

> The string `${filename}` is automatically replaced with the name of the file provided by the user


```html
<form action="http://storage.googleapis.com" method="post" enctype="multipart/form-data">
  <input type="hidden" name="key" value="postobjecttest2019110702/${filename}">
  <input type="hidden" name="GoogleAccessId" value="...">
  <input type="hidden" name="policy" value="...">
  <input type="hidden" name="signature"
         value="...">
  <input name="file" type="file">
  <input type="submit" value="Upload">
</form>
```

Reply:
I misread the documentation. Thanks for clarifying @quartzmo! 

Approving, let's merge this in.